_drafts/gnu-make-guidelines.md

```diff
@@ -0,0 +1,845 @@
+
+```makefile
+foo.txt:
+	# a simple recipe
+	echo HEYO > $@
+```
+
+If any line of the recipe returns a non-zero exit code, Make will terminate and
+print an error message. You can tell Make to ignore non-zero exit codes by
+prefixing with the `-` character:
+
+```makefile
+.PHONY: clean
+clean:
+	# we don't care if rm fails
+	-rm -f ./build
+```
+
+Prefixing a recipe line with `@` will disable echoing that line before
+executing:
+
+```makefile
+clean:
+	@# this recipe will just print 'About to clean everything!'
+	@# prefixing the shell comment lines '#' here also prevents them from
+	@# appearing during execution
+	@echo About to clean everything!
+```
+
+Make will expand variable/function expressions in the recipe context before
+running them, but otherwise not process it. If you want to access shell
+variables, escape them with `$`:
+
+```makefile
+USER = linus
+
+print-user:
+	# print out the shell variable $USER
+	echo $$USER
+
+	# print out the make variable USER
+	echo $(USER)
+```
+
+## Functions
+
+Make functions are called with the syntax:
+
+```makefile
+$(function-name arguments)
+```
+
+`arguments` is a comma-delimited list of arguments.
+
+### Built in functions
+
+There are several functions provided by Make. I most commonly use those for text
+manipulation:
+[https://www.gnu.org/software/make/manual/html_node/Text-Functions.html](https://www.gnu.org/software/make/manual/html_node/Text-Functions.html)
+[https://www.gnu.org/software/make/manual/html_node/File-Name-Functions.html](https://www.gnu.org/software/make/manual/html_node/File-Name-Functions.html)
+
+For example:
+
+```makefile
+FILES=$(wildcard *.c)
+
+# you can combine function calls; here we strip the suffix off of $(FILES) with
+# the $(basename) function, then add the .o suffix
+O_FILES=$(addsuffix .o,$(basename $(FILES)))
+
+# note that the GNU Make Manual suggests an alternate form for this particular
+# operation:
+O_FILES=$(FILES:.c=.o)
+```
+
+### User-defined functions
+
+You can define your own functions as well:
+
+```makefile
+reverse = $(2) $(1)
+
+foo = $(call reverse,a,b)
+```
+
+A more complicated but quite useful example:
+
+```makefile
+# recursive wildcard (use it instead of $(shell find . -name '*.c'))
+# taken from https://stackoverflow.com/a/18258352
+rwildcard=$(foreach d,$(wildcard $1*),$(call rwildcard,$d/,$2) $(filter $(subst *,%,$2),$d))
+
+C_FILES = $(call rwildcard,.,*.c)
+```
+
+### Shell function
+
+You can have Make call a shell expression and capture the result:
+
+```makefile
+TODAYS_DATE=$(shell date --iso-8601)
+```
+
+I'm cautious when using this feature, though; it adds a dependency on whatever
+programs you use, so if you're calling more exotic programs, make sure your
+build environment is controlled (eg in a container or with
+[Conda](https://interrupt.memfault.com/blog/conda-developer-environments)).
+
+## Conditionals
+
+Make has syntax for conditional expressions:
+
+```makefile
+FOO=yolo
+ifeq ($(FOO),yolo)
```

Comment:
Unfortunately, the syntax highlighting in Jekyll does not like Makefile syntax. For the others, it isnt' so bad but this one is really bad. I'd probably remove the `makefile` language specification